### ISSUE TITLE
Audible quality loss in transcoded uploads

### ISSUE THREAD

**Issue by SharkWipf** [thebigmunch/google-music-scripts#8]
- `gms -v` : `google-music-scripts 2.0.0`.
- Installed through `pip install google-music-scripts` in a clean python3.6 virtualenv.

When running `gms upload somesong.flac`, the resulting uploaded song will be of really really bad quality.  
When manually converting the song to mp3 first (`ffmpeg -i in.flac -ab 320k out.mp3`) and then uploading through gms, the song sounds fine.  
Adding the --no-transcode-lossless flag causes the file to become unplayable when uploaded (probably to be expected, but might be worth explaining in the --help output).  
Adding the --transcode-lossy flag and uploading an mp3 does not result in notably worse quality, so it seems to be lossless files specific.

**Comment by thebigmunch** [comment]
Yeah, I'm going to just remove direct FLAC uploading. It's awfully fussy, and you can't stream or download the FLAC file anyway.

But still looking to get more info on the other issues.

**Comment by SharkWipf** [comment]
> I'm not sure what you mean by really bad quality. Send me some files and tell me what I'm listening for?

I've uploaded a snippet of a song [before](https://sharkwipf.nl/01%20-%20only%20my%20railgun%20sample%20orig.mp3) and [after](https://sharkwipf.nl/01%20-%20only%20my%20railgun%20sample%20fried.mp3) using `gsm` to upload it to Gmusic. The latter was downloaded through the Gmusic web interface.  
I don't think I need to clarify what to listen for.  

> I'd also try comparing the result of uploading with Google's Music Manager.

I would love to, but I'm using `gsm` because I can't.  For reasons unknown, every time I try to upload it throws an error. 
![image](https://user-images.githubusercontent.com/751205/49259450-b31aee80-f439-11e8-89be-8f0459dd0e29.png)

Reportedly this error has to do with either duplicate mac addresses or not having an eth0 on Linux, sadly I have some bridges that have linked mac addresses and "predictable interface names" has renamed my eth0 to `enp0s31f6`, so I cannot use Google Music Manager on this machine at all.

> I also am not having any problem playing back MP3s uploading using google-music-scripts on Android.

I used the now-deprecated `gmusicapi-scripts` for this and didn't have any issues, but it's been a while since I had to upload music myself (thanks, region locking).  
Now I'm just getting a generic "Music playback error: Couldn't play the track you requested" on anything I've uploaded through `gms`.

> Perhaps you could try replacing your version of ffmpeg and see what happens?

That shouldn't be making a difference I'd think, as converting through the same ffmpeg by hand does create a proper sounding file (albeit still unplayable on Android when uploaded).  
I'll try ffmpeg git, 4.0 and 3.4.5 though (currently on ffmpeg 4.1).

**Comment by SharkWipf** [comment]
Update: I've tested all of above mentioned ffmpeg versions, with no difference in result.  
Afterwards I decided to try out the old `gmupload` from `gmusicapi-scripts` in a clean virtualenv and, to my pleasant surprise, that one did still work perfectly fine, gets uploaded in proper quality and plays fine on Android too.

**Comment by thebigmunch** [comment]
Yeah, I'd like to get at least a file or two that exhibit this for you so I can test them out specifically. That's why I asked for them; my email address is in my profile.

And just to clarify, these were uploading, not matched, right? One of the major differences between my libraries and gmusicapi is that gmusicapi has never calculated the ``clientId`` properly due to mutagen not being sufficient for the most part. So files would only get matched if there was a metadata match but not a ``clientId`` match. That's why I don't have an option for disabling matching yet (possibly never). It's actually a lot more difficult to spoof when you're calculating the ``clientId`` correctly, and it might not be possible to spoof metadata-only matching. That's one of the reasons I was interested in any differences between Google's Music Manager and this.

**Comment by thebigmunch** [comment]
OK. So, the benefit of taking a nice fresh look at things is spotting things you missed the first time around. There **is** in fact a difference in the transcoding command from previously. I forgot to add an if check that I removed when I was testing something way back when *facepalm*. 

Thanks for finding this. I haven't been uploading and listening much recently, so it might've been a while before I noticed myself. So, that'll get fixed in the upstream library.

**Comment by thebigmunch** [comment]
Alright, another upgrade install should get you ``google-music-proto==2.1.2`` which contains the transcoding fix.

**Comment by SharkWipf** [comment]
Sorry for the delay, I didn't have time to test the new changes out yet.  
I just tried it and this issue seems fine now, FLAC files are uploaded in proper quality *and* they work fine on Android now.  
Thanks for the fix!

**Comment by thebigmunch** [comment]
Gonna leave this open for a bit for visibility.
